esi/lib/nodes.py

```diff
@@ -0,0 +1,110 @@
+
+#   Licensed under the Apache License, Version 2.0 (the "License"); you may
+#   not use this file except in compliance with the License. You may obtain
+#   a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+
+import concurrent.futures
+
+from esi.lib import networks
+
+
+def node_and_port_list(connection, filter_node=None):
+    nodes = None
+    ports = None
+
+    if filter_node:
+        nodes = [connection.baremetal.find_node(name_or_id=filter_node,
+                                                ignore_missing=False)]
+        ports = connection.baremetal.ports(details=True, node_id=nodes[0].id)
+    else:
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            f1 = executor.submit(connection.baremetal.nodes)
+            f2 = executor.submit(connection.baremetal.ports, details=True)
+            nodes = list(f1.result())
+            ports = list(f2.result())
+
+    return nodes, ports
+
+
+"""
++--------------+-------------------+-------------------------------+--------------------+----------------+------------------+---------------+
+| Node         | MAC Address       | Port                          | Network            | Fixed IP       | Floating Network | Floating IP   |
++--------------+-------------------+-------------------------------+--------------------+----------------+------------------+---------------+
+| oct4-cha3-1b | a8:99:69:a7:0a:ef | esi-oct4-cha3-1b-provisioning | provisioning (623) | 192.168.11.138 | external         | 128.31.20.216 |
++--------------+-------------------+-------------------------------+--------------------+----------------+------------------+---------------+
+"""
+
+
+def network_list(connection, filter_node=None, filter_network=None):
+    """List nodes and their network attributes
+
+    :param connection: An OpenStack connection
+    :type connection: :class:`~openstack.connection.Connection`
+    :param filter_node: Name or UUID of the node
+    :param filter_network: The name or ID of a network
+    """
```

Comment:
I like this - these functions are complex, and knowing what to expect is a good idea. Can you also document what it returns, with an example? That would help anyone else using the function. It makes sense to add similar documentation to the other functions here and in networks.py